### PR TITLE
doctest addons

### LINKED ISSUE #582: Build doctest to test python docstrings
Seems that core TF has already used `doctest` to test python docstring, which having different style with ours, especially for code snippet example usage. To unify the docstring and for better documentation, the doctest should be also built in addons.

See https://www.tensorflow.org/community/contribute/docs_ref for more details.

### PR DESCRIPTION
**Pull request by autoih**
Address [issue 582](https://github.com/tensorflow/addons/issues/582). Use DocTest to test the Python docstrings.

EDIT (@gabrieldemarmiesse) : fix #582

### CODE REVIEW

File: tools/docs/tfa_doctest.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    def check_output(self, want, got, optionflags):
+        # Replace python's addresses with ellipsis (`...`) since it can change on
+        # each execution.
+        want = self.ADDRESS_RE.sub('at ...>', want)
+        return doctest.OutputChecker.check_output(self, want, got, optionflags)
+
+    _MESSAGE = textwrap.dedent("""\n
+        #############################################################
+        Check the documentation
+        (https://www.tensorflow.org/community/contribute/docs_ref) on how to write testable docstrings.
```
Comment (autoih) [inline tools/docs/tfa_doctest.py]: Should we use the same [link](https://www.tensorflow.org/community/contribute/docs_ref) or create a new one at somewhere?

Comment (WindQAQ) [reply tools/docs/tfa_doctest.py]: Use the same one is fine :-)

**Comment by seanpmorgan** [comment]
> @seanpmorgan Hi Sean, I wonder if we should also put the doctest on CI. My initial thought is not to do that. Maybe a neat description to ask contributors to so is enough. What do you think about it? Thank you!
> 
> @autoih Thanks for the contribution! Would you mind also making other code snippets compatible with doctest?

Hmmmm I think once we fully enforce this it could be a good idea to include in CI (depending on the runtime). But for the short-term I'm fine with this being a user ran test.

**Comment by WindQAQ** [comment]
> Hmmmm I think once we fully enforce this it could be a good idea to include in CI (depending on the runtime). But for the short-term I'm fine with this being a user ran test.

@autoih Could you share the runtime for the whole repo?

**Comment by autoih** [comment]
Thanks @WindQAQ and @seanpmorgan. The runtime of the 11 files is about 2.482 seconds.

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent.

**Comment by gabrieldemarmiesse** [comment]
Thanks a lot @autoih for your work! I tried to symplify the code and I added the test in the CI. 

There was a code snippet which was wrong so I didn't fix it (outside the scope of this pull request). I added a TODO to avoid forgetting about it.

This pull request is ready for review again :)

**Comment by autoih** [comment]
> There was a code snippet which was wrong so I didn't fix it (outside the scope of this pull request). I added a TODO to avoid forgetting about it.
Thanks @gabrieldemarmiesse, can you please point it out where it is?

**Comment by gabrieldemarmiesse** [comment]
I'm afraid you may have overriten my implementation when you forced-push, it's no big deal, I can do it again, it's pretty easy with pytest. Or maybe you'd wish to try? http://doc.pytest.org/en/latest/doctest.html 

We don't actually need the whole TF/bazel/absl machinery to test the docstrings.

**Comment by autoih** [comment]
Thanks @gabrieldemarmiesse. Let me do it first, and may need your help/review later.  :smiley:

**Comment by autoih** [comment]
@gabrieldemarmiesse, since using `pytest --doctest-glob='*_test.py'` will be run through all `doctest`, I'd like to ask your guidance for this PR (e.g,. should I remove `docs/tfa_doctest.py`, etc?). Can you plesae give me some suggestions? Thanks.

**Comment by gabrieldemarmiesse** [comment]
Yeah I don't think we need this file. Maybe just add an entry to the sanity check dockerfile and run it in the CI?

**Comment by autoih** [comment]
Thanks @gabrieldemarmiesse, my pc encounters some issues, may need time to fix, can you please help the PR, Dockerfile? Many thanks.

**Comment by gabrieldemarmiesse** [comment]
Sure thing. I'm unsure of the timeline since I spend a lot of time migrating the tests suite to pytest.

**Comment by autoih** [comment]
Thanks @gabrieldemarmiesse, of course, I'll fix it once the issues fixed.

**Comment by autoih** [comment]
Thanks, @gabrieldemarmiesse, I fixed my machine issues. So, I simply add `RUN pytest --ignore-glob=*_test.py`  inside of `test the editable mode` in sanity.dockerfile, please correct it if it's inappropriate. Thanks.

**Comment by gabrieldemarmiesse** [comment]
Looks great, thanks!

Three small points:

*  Could you add the `-v` option for pytest to print the tests list? Also are you sure that it actually runs the doctests? I would expect pytest to take the argument `--doctest-modules` for that. You might also have to use fixtures in the conftest.py to allow the usage of `tf`, `tfa` and `np`. See [this section in the docs](http://doc.pytest.org/en/latest/doctest.html#doctest-namespace-fixture). 
* You should remove the two unused empty files and the section you added in the `BUILD` file. See the section [Files changed](https://github.com/tensorflow/addons/pull/677/files) of your pull request.

**Comment by autoih** [comment]
Thanks for the suggestions, @gabrieldemarmiesse.  This pr may not consider test files for doctest.

**Comment by autoih** [comment]
@gabrieldemarmiesse, can you please help review when you have time? Thanks.

File: tensorflow_addons/metrics/cohens_kappa.py [hunk 2]
```diff
@@ -27,34 +28,26 @@
...
     print('Final result: ', m.result().numpy()) # Result: 0.61904764
-
     # To use this with weights, sample_weight argument can be used.
     m = tfa.metrics.CohenKappa(num_classes=5)
     m.update_state(actuals, preds, sample_weight=weights)
     print('Final result: ', m.result().numpy()) # Result: 0.37209308
     ```
-
     Usage with tf.keras API:
-
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/cohens_kappa.py:57]: Why this is being modified?

File: tensorflow_addons/metrics/multilabel_confusion_matrix.py [hunk 3]
```diff
@@ -45,30 +46,6 @@ class MultiLabelConfusionMatrix(Metric):
...
-    y_true = tf.constant([[1, 0, 0], [0, 1, 0]],
-             dtype=tf.int32)
-    y_pred = tf.constant([[1, 0, 0],[0, 0, 1]],
-             dtype=tf.int32)
-    output = MultiLabelConfusionMatrix(num_classes=3)
-    output.update_state(y_true, y_pred)
-    print('Confusion matrix:', output.result().numpy())
-
-    # Confusion matrix: [[[1 0] [0 1]] [[1 0] [1 0]] [[1 1] [0 0]]]
-    ```
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/multilabel_confusion_matrix.py:71]: Why the example is removed here? Maybe also change this to
```
Usage: 

>>> xxx
```

File: tensorflow_addons/conftest.py [hunk 4]
```diff
@@ -5,6 +5,28 @@
...
+
+
+@pytest.fixture(autouse=True)
+def add_tf(doctest_namespace):
+    doctest_namespace["tf"] = tf
+
+
+@pytest.fixture(autouse=True)
+def add_tfa(doctest_namespace):
+    doctest_namespace["tfa"] = tfa
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/conftest.py:32]: You can use one function for all that.

File: tools/docker/sanity_check.Dockerfile [hunk 5]
```diff
@@ -118,6 +118,7 @@ RUN --mount=type=cache,id=cache_bazel,target=/root/.cache/bazel \
     bash tools/install_so_files.sh
 RUN pip install --no-deps -e .
 RUN pytest -v -n auto ./tensorflow_addons/activations
+RUN pytest -v --doctest-modules tensorflow_addons/ --ignore-glob=*_test.py
```
Comment (gabrieldemarmiesse) [inline tools/docker/sanity_check.Dockerfile:121]: Is it possible to group everything in one call?

File: tensorflow_addons/metrics/cohens_kappa.py (hunk 2 above)
Comment (autoih) [reply tensorflow_addons/metrics/cohens_kappa.py:57]: I temporarily use the original docstr, since there may be something wrong of the example, please see [here](https://colab.research.google.com/drive/1A7KZ27PRtY-nBMpG6oF_RSCMm7Tjs3zM#scrollTo=hfBtoG7T6Nc6)

File: tensorflow_addons/metrics/multilabel_confusion_matrix.py (hunk 3 above)
Comment (autoih) [reply tensorflow_addons/metrics/multilabel_confusion_matrix.py:71]: Thanks @WindQAQ. Same issue to cohens_kappa. May I have your suggestion?

File: tools/docker/sanity_check.Dockerfile (hunk 5 above)
Comment (autoih) [reply tools/docker/sanity_check.Dockerfile]: Thanks @gabrieldemarmiesse.

File: tensorflow_addons/metrics/cohens_kappa.py (hunk 2 above)
Comment (autoih) [reply tensorflow_addons/metrics/cohens_kappa.py:57]: Fixed it, the problem is from original code.

File: tensorflow_addons/metrics/multilabel_confusion_matrix.py (hunk 3 above)
Comment (autoih) [reply tensorflow_addons/metrics/multilabel_confusion_matrix.py:71]: Thanks @WindQAQ. Fixed.

**Comment by autoih** [comment]
Thanks for the comments @gabrieldemarmiesse. Not sure if macos will be an issue.

**Comment by autoih** [comment]
@gabrieldemarmiesse, any other suggestions?

File: tensorflow_addons/losses/focal_loss.py [hunk 6]
```diff
@@ -37,15 +38,11 @@ class SigmoidFocalCrossEntropy(LossFunctionWrapper):
...
-      [[0.97], [0.91], [0.03]],
-      [[1.0], [1.0], [0.0]])
-    print('Loss: ', loss.numpy())  # Loss: [0.00010971,
-                                            0.0032975,
-                                            0.00030611]
-    ```
+    >>> fl = tfa.losses.SigmoidFocalCrossEntropy()
+    >>> loss = fl([[0.97], [0.91], [0.03]], [[1.0], [1.0], [0.0]])
+    >>> loss
+    <tf.Tensor: shape=(3,), dtype=float32, numpy=array([0.00010971, 0.00329749, 0.00030611], dtype=float32)>
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/losses/focal_loss.py:44]:
```suggestion
    <tf.Tensor: shape=(3,), dtype=float32, numpy=array([10971, 0.00329749, 0.00030611], dtype=float32)>
```

**Comment by gabrieldemarmiesse** [comment]
I modified a line to make sure the tests for dostrings were actually running in the CI.

**Comment by gabrieldemarmiesse** [comment]
@autoih it seems the tests were not running. I changed a number and the CI is still passing. Could you make sure the tests are running correctly? You can ping me afterwards and I'll review.

**Comment by autoih** [comment]
Thanks @gabrieldemarmiesse. For some reason, `--doctest-modules` is missing in sanity check dockerfile, put it back now.

**Comment by autoih** [comment]
@googlebot I fixed it.
